File: src/leetcode/2026/04/Solution804.java
import java.util.HashSet;
import java.util.Set;

public class Solution804 {
    public int uniqueMorseRepresentations(String[] words) {
        String[] morse = { ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--",
                "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.." };

        Set<String> uniqueTransformations = new HashSet<>();
        for (String word : words) {
            StringBuilder transformation = new StringBuilder();
            for (char c : word.toCharArray()) {
                transformation.append(morse[c - 'a']);
            }
            uniqueTransformations.add(transformation.toString());
        }
        return uniqueTransformations.size();
    }

    public static void main(String[] args) {
        Solution804 solution = new Solution804();

        String[] words1 = { "gin", "zen", "gig", "msg" };
        System.out.println(solution.uniqueMorseRepresentations(words1)); // Output: 2

        String[] words2 = { "a" };
        System.out.println(solution.uniqueMorseRepresentations(words2)); // Output: 1
    }
}

File: src/leetcode/2026/04/Solution804.cpp
#include <string>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution804 {
public:
    int uniqueMorseRepresentations(vector<string>& words) {
        string morse[26] = { ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--",
                             "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--",
                             "--.." };
        unordered_set<string> uniqueTransformations;
        for (const string& word : words) {
            string transformation;
            for (char c : word) {
                transformation += morse[c - 'a'];
            }
            uniqueTransformations.insert(transformation);
        }
        return uniqueTransformations.size();
    }
};

int main() {
    Solution804 solution;

    vector<string> words1 = { "gin", "zen", "gig", "msg" };
    printf("Unique Morse representations: %d\n", solution.uniqueMorseRepresentations(words1)); // Output: 2

    vector<string> words2 = { "a" };
    printf("Unique Morse representations: %d\n", solution.uniqueMorseRepresentations(words2)); // Output: 1

    return 0;
}